compiler-plugin/src/main/java/io/ballerina/stdlib/email/compiler/EmailServiceValidator.java

```diff
@@ -248,13 +251,21 @@ private boolean hasNoParameters(SeparatedNodeList<ParameterNode> parameterNodes,
         return false;
     }
 
-    private void validateParameter(SeparatedNodeList<ParameterNode> parameterNodes, String functionName) {
+    private void validateParameter(FunctionDefinitionNode functionDefinitionNode,
+            SeparatedNodeList<ParameterNode> parameterNodes, String functionName) {
+        FunctionTypeSymbol functionTypeSymbol = ((MethodSymbol) ctx.semanticModel().symbol(functionDefinitionNode)
+                .get()).typeDescriptor();
+        List<ParameterSymbol> inputParams = functionTypeSymbol.params().get();
+        String moduleId = getModuleId(inputParams.get(0));
+
         for (ParameterNode parameterNode : parameterNodes) {
             RequiredParameterNode requiredParameterNode = (RequiredParameterNode) parameterNode;
             Node parameterTypeName = requiredParameterNode.typeName();
             DiagnosticInfo diagnosticInfo;
-            if (functionName.equals(ON_MESSAGE) && (!parameterTypeName.toString().contains(EMAIL_MESSAGE)
-                    || !parameterTypeName.toString().contains(MODULE_NAME))) {
+            if (functionName.equals(ON_MESSAGE) && !(inputParams.get(0).typeDescriptor().signature().
+                    equals(moduleId + ":" + EMAIL_MESSAGE) || (inputParams.get(0).typeDescriptor().signature().
+                    contains(EMAIL_MESSAGE) && inputParams.get(0).typeDescriptor().typeKind() ==
+                    TypeDescKind.INTERSECTION))) {
```

Comment:
Shall we extract this validation into a another method ?

Reply:
Extracted.